fix: git#865 - fixes notifying the switch atSign event multiple times

sitaram-kalluri opened the pull request:
**- What I did**
* Remove the inactive listeners from `AtClientManager._changeListeners` list.
* Add unit tests to assert the changes.
* (gkc) Some further simplifications to code associated with switching atSigns

**- How I did it**
* Added `removeChangeListener` method in AtClientManager, that removes the given instance of AtSignChangeListener from `AtClientManager._changeListeners` list
* In `AtClientManager._notifyListeners` added a line that clones the _changeListeners list and iterate on the cloned list to avoid the concurrent modification exception.
* Some changes to simplify code
    * fix: Added check to SwitchAtSignEvent constructor to prevent events being created when the atClients are the same
    * debug: Added a logger to AtClientManager. Log INFO when switching atSigns
    * refactor: Removed instance variable _previousAtClient from AtClientManager. Instead, determine previousAtClient within the setCurrentAtSign method
    * fix: Added duplicate check to AtClientManager.listenToAtSignChange
    * fix: Changed _currentAtClient from `late` to being nullable to prevent uninitialized check in setCurrentAtSign
    * fix: AtClientManager now calls _notifyListeners before creating the new NotificationServiceImpl and SyncServiceImpl. This allows those classes' implementations of listenToAtSignChange to be simplified
    * refactor: Removed static maps of instances in NotificationServiceImpl and SyncServiceImpl thus allowing some code simplification
    * refactor: Simplified NotificationServiceImpl.listenToAtSignChange
    * refactor: Simplified SyncServiceImpl.listenToAtSignChange, and added code to close the RemoteSecondary's AtLookUp's connection if there is one

**- How to verify it**
* On the switch atSign event, `AtClientManager._changeListeners` should not have obsolete listeners.

**- Description for the changelog**
* Remove the inactive listeners from `AtClientManager._changeListeners` list.

<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

sitaram-kalluri committed:
fix: git#865 - fixes notifying the switch atSign event multiple times

sitaram-kalluri commented:
Found the root cause of failing end2end_tests. Will fix the issue and push the changes.

gkc reviewed (dismissed):
As discussed just now, we don't need the 'isActive' flag. The concurrent modification error can be eliminated by having `_notifyListeners` make a copy of the _changeListeners and iterate through the copy of the list rather than through the list itself.

As a side note - force-pushing your branch is masking useful information about the evolution of the code - without seeing the interim commits, it was completely unclear to me why the isActive flag was added until we had our discussion. Is there a reason you prefer force-push?

sitaram-kalluri commented:
> As discussed just now, we don't need the 'isActive' flag. The concurrent modification error can be eliminated by having `_notifyListeners` make a copy of the _changeListeners and iterate through the copy of the list rather than through the list itself.
> 
> As a side note - force-pushing your branch is masking useful information about the evolution of the code - without seeing the interim commits, it was completely unclear to me why the isActive flag was added until we had our discussion. Is there a reason you prefer force-push?

@gkc: Sure, will update the code as per our discussion. I generally do a force push to keep the commit history clean. However, I do it only when there are no review comments/discussions on the PR. If there are review comments/discussions on the PR I avoid doing a force push.

gkc commented:
> @gkc: Sure, will update the code as per our discussion. I generally do a force push to keep the commit history clean. However, I do it only when there are no review comments/discussions on the PR. If there are review comments/discussions on the PR I avoid doing a force push.

I understand. However sometimes (as in this case), squashing the commits makes it impossible to understand how you got to a particular design / implementation decision, without having a conversation.

sitaram-kalluri commented:
> > @gkc: Sure, will update the code as per our discussion. I generally do a force push to keep the commit history clean. However, I do it only when there are no review comments/discussions on the PR. If there are review comments/discussions on the PR I avoid doing a force push.
> 
> I understand. However sometimes (as in this case), squashing the commits makes it impossible to understand how you got to a particular design / implementation decision, without having a conversation.

Ok Gary, Will avoid force pushing to the branch.

sitaram-kalluri committed:
fix: Remove isActive flag and remove inactive change listeners

sitaram-kalluri committed:
fix: Remove isActive flag from KeyStreamMixin

sitaram-kalluri committed:
Merge branch 'trunk' into unregister_change_listeners

sitaram-kalluri committed:
fix: Rename the variable in notifyListener method

VJag commented on packages/at_client/lib/src/manager/at_client_manager.dart:
```diff
@@ -74,8 +75,26 @@ class AtClientManager {
 
   void _notifyListeners(SwitchAtSignEvent switchAtSignEvent) {
-    for (var listener in _changeListeners) {
+    List<AtSignChangeListener> copyOfChangeListeners = List.from(_changeListeners);
```
Lets add Dart docs for the reason to clone the list

sitaram-kalluri replied:
Sure Jagan, Will add the comments

VJag commented on packages/at_client/lib/src/manager/at_client_manager.dart:
```diff
@@ -74,8 +75,26 @@ class AtClientManager {
     }
   }
+
+  /// Clears the inactive AtSignChangeListener from the _changeListeners list.
```
Dart doc should probably say:

Removes the given listener from the list of listeners, that are notified whenever an atSign is switched.

sitaram-kalluri replied:
Sure Jagan, Will update the dart docs

VJag commented on packages/at_client/lib/src/service/notification_service_impl.dart:
```diff
@@ -501,6 +504,9 @@ class NotificationServiceImpl
       stopAllSubscriptions();
       _logger.finer(
           'removing from _notificationServiceMap: ${_atClient.getCurrentAtSign()}');
+      _atClientManager.removeChangeListeners(
```
Unable to comment on an if condition above, just for my understanding, why is the if checking for previous and present atSign to be same?

if (switchAtSignEvent.previousAtClient?.getCurrentAtSign() ==
        _atClient.getCurrentAtSign()) {

sitaram-kalluri replied:
In the above IF condition, _atClient.getCurrentAtSign returns the atSign with which the `NotificationServiceImpl` is created. Since we only wanted to remove the previous atSign (not the new atSign) added the above IF condition.

VJag replied:
Understood the code. Pls add dart docs if it makes sense.

sitaram-kalluri committed:
fix: add unit tests and dart docs

sitaram-kalluri committed:
Merge branch 'trunk' into unregister_change_listeners

VJag commented on packages/at_client/lib/src/manager/at_client_manager.dart:
```diff
@@ -74,8 +75,29 @@ class AtClientManager {
   }
+
+  /// Removes the given listener from the _changeListeners list.
+  void removeChangeListeners(AtSignChangeListener atSignChangeListener) {
```
can you change Dart docs to: "Removes the given listener from the list of listeners, that are notified whenever the @sign is switched".

VJag replied:
general rule: No impl details in Dart docs

sitaram-kalluri replied:
Sure Jagan, will update the dart docs

sitaram-kalluri committed:
Merge branch 'trunk' into unregister_change_listeners

sitaram-kalluri committed:
fix: update the dart docs

gkc committed:
fix: Added check to SwitchAtSignEvent constructor to prevent events being created when the atClients are the same
debug: Added a logger to AtClientManager. Log INFO when switching atSigns
refactor: Removed instance variable _previousAtClient from AtClientManager. Instead, determine previousAtClient within the setCurrentAtSign method
fix: Added duplicate check to AtClientManager.listenToAtSignChange
fix: Changed _currentAtClient from `late` to being nullable to prevent uninitialized check in setCurrentAtSign
fix: AtClientManager now calls _notifyListeners before creating the new NotificationServiceImpl and SyncServiceImpl
refactor: Removed static maps of instances in NotificationServiceImpl and SyncServiceImpl thus allowing some code simplification
refactor: Simplified NotificationServiceImpl.listenToAtSignChange
refactor: Simplified SyncServiceImpl.listenToAtSignChange, and added code to close the RemoteSecondary's AtLookUp's connection if there is one

gkc committed:
style: Removed commented-out code

gkc commented:
@VJag Have made some further changes to this branch in https://github.com/atsign-foundation/at_client_sdk/pull/867/commits/0db8b2b8c93708b3459651b4db54ccd718cdb9dd ... please review

gkc committed:
Merge branch 'trunk' into unregister_change_listeners

sitaram-kalluri committed:
fix: remove connection.close in SyncServiceImpl.dart
